core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -473,6 +473,324 @@ class DAGSchedulerSuite
     assertDataStructuresEmpty()
   }
 
+  // Helper function to validate state when creating tests for task failures
+  def checkStageId(stageId: Int, attempt: Int, stageAttempt: TaskSet) {
+    assert(stageAttempt.stageId === stageId)
+    assert(stageAttempt.stageAttemptId == attempt)
+  }
+
+  def makeCompletions(stageAttempt: TaskSet, reduceParts: Int): Seq[(Success.type, MapStatus)] = {
+    stageAttempt.tasks.zipWithIndex.map { case (task, idx) =>
+      (Success, makeMapStatus("host" + ('A' + idx).toChar, reduceParts))
+    }.toSeq
+  }
+
+  def setupStageAbortTest(sc: SparkContext) {
+    sc.listenerBus.addListener(new EndListener())
+    ended = false
+    jobResult = null
+  }
+
+  // Create a new Listener to confirm that the listenerBus sees the JobEnd message
+  // when we abort the stage. This message will also be consumed by the EventLoggingListener
+  // so this will propagate up to the user.
+  var ended = false
+  var jobResult : JobResult = null
+
+  class EndListener extends SparkListener {
+    override def onJobEnd(jobEnd: SparkListenerJobEnd): Unit = {
+      jobResult = jobEnd.jobResult
+      ended = true
+    }
+  }
+
+  /**
+   * In this test we simulate a job failure where the first stage completes successfully and
+   * the second stage fails due to a fetch failure. Multiple successive fetch failures of a stage
+   * trigger an overall stage abort to avoid endless retries.
+   */
+  test("Multiple consecutive stage failures should lead to task being aborted.") {
+    setupStageAbortTest(sc)
+
+    val shuffleMapRdd = new MyRDD(sc, 2, Nil)
+    val shuffleDep = new ShuffleDependency(shuffleMapRdd, null)
+    val shuffleId = shuffleDep.shuffleId
+    val reduceRdd = new MyRDD(sc, 2, List(shuffleDep))
+    submit(reduceRdd, Array(0, 1))
+
+    for (attempt <- 0 until Stage.MAX_STAGE_FAILURES) {
+      // Complete all the tasks for the current attempt of stage 0 successfully
+      val stage0Attempt = taskSets.last
+      checkStageId(0, attempt, stage0Attempt)
+
+      // Run stage 0
+      complete(stage0Attempt, makeCompletions(stage0Attempt, 2))
```

Comment:
Hi @ilganeli after taking another look at this code after a break, I think we can clean up the test cases a bit to make it a little easier to follow.  In all of these tests, we repeat a common pattern of:

1) grab the last taskSet
2) check its the stage & attempt we think it is
3) send it some task completions:
3a) successfully complete a ShuffleMapStage with the right MapStatus
3b) Fail one task with a fetch failure
3c) successfully complete all result tasks with "42"

We can make a set of util methods to do all of that, which would reduce a lot of the noise in these tests, eg. `completeNextShuffleMapSuccessfully(stageId: Int, stageAttemptId:Int)`

In fact, you could also eliminate this annoying detail of having to know the number of partitions for the stage that reads the shuffle map output, since you can grab that with `scheduler.stageIdToStage(stageId).asInstanceOf[ShuffleMapStage].shuffleDep.partitioner.numPartitions`